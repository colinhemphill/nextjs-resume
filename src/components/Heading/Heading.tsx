import { VariantProps, cva } from 'class-variance-authority';
import { DetailedHTMLProps, HTMLAttributes, PropsWithChildren } from 'react';
import { twMerge } from 'tailwind-merge';

const headingVariants = cva('font-bold', {
  variants: {
    color: {
      danger: 'text-danger-11',
      neutral: 'text-neutral-12',
      neutralSubtle: 'text-neutral-11',
      primary: 'text-primary-11',
    },
    size: {
      1: 'text-4xl md:text-5xl',
      2: 'text-2xl md:text-3xl',
      3: 'text-xl md:text-2xl',
      4: 'text-lg md:text-xl',
      5: 'text-base md:text-lg',
      6: 'text-sm md:text-base',
    },
  },
  defaultVariants: {
    color: 'neutral',
  },
});

export type HeadingProps = DetailedHTMLProps<
  HTMLAttributes<HTMLHeadingElement>,
  HTMLHeadingElement
> &
  VariantProps<typeof headingVariants> & {
    level: 1 | 2 | 3 | 4 | 5 | 6;
  };

export function Heading({
  children,
  className,
  color,
  level,
  size,
}: PropsWithChildren<HeadingProps>) {
  const classes = twMerge(
    headingVariants({ color, size: size || level }),
    className,
  );

  switch (level) {
    case 1: {
      return <h1 className={classes}>{children}</h1>;
    }
    case 2: {
      return <h2 className={classes}>{children}</h2>;
    }
    case 3: {
      return <h3 className={classes}>{children}</h3>;
    }
    case 4: {
      return <h4 className={classes}>{children}</h4>;
    }
    case 5: {
      return <h5 className={classes}>{children}</h5>;
    }
    case 6: {
      return <h6 className={classes}>{children}</h6>;
    }
  }
}
