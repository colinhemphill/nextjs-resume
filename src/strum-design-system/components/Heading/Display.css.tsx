import { style, styleVariants } from '@vanilla-extract/css';
import { vars } from '../../themes/contract.css';
import { breakpointMediaQueries } from '../../utils/breakpointMediaQueries';

const displayBaseStyle = style({
  fontWeight: 700,
  lineHeight: 1.3,
  marginBottom: vars.spacers[2],
  marginTop: 0,
});

export const displayStyleVariants = styleVariants({
  1: [
    displayBaseStyle,
    {
      '@media': {
        [breakpointMediaQueries.xl]: {
          fontSize: '5rem',
        },
      },
      fontSize: 'calc(1.625rem + 4.5vw)',
    },
  ],
  2: [
    displayBaseStyle,
    {
      '@media': {
        [breakpointMediaQueries.xl]: {
          fontSize: '4.5rem',
        },
      },
      fontSize: 'calc(1.575rem + 3.9vw)',
    },
  ],
  3: [
    displayBaseStyle,
    {
      '@media': {
        [breakpointMediaQueries.xl]: {
          fontSize: '4rem',
        },
      },
      fontSize: 'calc(1.525rem + 3.3vw)',
    },
  ],
  4: [
    displayBaseStyle,
    {
      '@media': {
        [breakpointMediaQueries.xl]: {
          fontSize: '3.5rem',
        },
      },
      fontSize: 'calc(1.475rem + 2.7vw)',
    },
  ],
  5: [
    displayBaseStyle,
    {
      '@media': {
        [breakpointMediaQueries.xl]: {
          fontSize: '3rem',
        },
      },
      fontSize: 'calc(1.425rem + 2.1vw)',
    },
  ],
  6: [
    displayBaseStyle,
    {
      '@media': {
        [breakpointMediaQueries.xl]: {
          fontSize: '2.5rem',
        },
      },
      fontSize: 'calc(1.375rem + 1.5vw)',
    },
  ],
});
