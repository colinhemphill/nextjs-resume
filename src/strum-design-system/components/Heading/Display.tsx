import React, { PropsWithChildren } from 'react';
import { Atoms } from '../../sprinkles.css';
import { composeWithAtoms } from '../../utils/compose';
import { displayStyleVariants } from './Display.css';

interface DisplayBase {
  atoms?: Atoms;
  level: 1 | 2 | 3 | 4 | 5 | 6;
  size: 1 | 2 | 3 | 4 | 5 | 6;
  text: string;
}

type Display = RequireOnlyOne<
  PropsWithChildren<DisplayBase>,
  'text' | 'children'
>;

const Display: React.FC<Display> = (props) => {
  const { atoms: atomicProperties, children, level, size, text } = props;

  const style = composeWithAtoms(atomicProperties, displayStyleVariants[size]);

  switch (level) {
    case 1:
      return <h1 className={style}>{text || children}</h1>;
    case 2:
      return <h2 className={style}>{text || children}</h2>;
    case 3:
      return <h3 className={style}>{text || children}</h3>;
    case 4:
      return <h4 className={style}>{text || children}</h4>;
    case 5:
      return <h5 className={style}>{text || children}</h5>;
    case 6:
      return <h6 className={style}>{text || children}</h6>;
  }
};

export default Display;
