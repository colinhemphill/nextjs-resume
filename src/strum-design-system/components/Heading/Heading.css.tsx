import { style, styleVariants } from '@vanilla-extract/css';
import { vars } from '../../themes/contract.css';
import { breakpointMediaQueries } from '../../utils/breakpointMediaQueries';

const headingBaseStyle = style({
  fontWeight: 700,
  lineHeight: 1.3,
  marginBottom: vars.spacers[2],
  marginTop: 0,
});

export const headingStyleVariants = styleVariants({
  1: [
    headingBaseStyle,
    {
      '@media': {
        [breakpointMediaQueries.xl]: {
          fontSize: '2.5rem',
        },
      },
      fontSize: 'calc(1.375rem + 1.5vw)',
    },
  ],
  2: [
    headingBaseStyle,
    {
      '@media': {
        [breakpointMediaQueries.xl]: {
          fontSize: '2rem',
        },
      },
      fontSize: 'calc(1.325rem + .9vw)',
    },
  ],
  3: [
    headingBaseStyle,
    {
      '@media': {
        [breakpointMediaQueries.xl]: {
          fontSize: '1.75rem',
        },
      },
      fontSize: 'calc(1.3rem + .6vw)',
    },
  ],
  4: [
    headingBaseStyle,
    {
      '@media': {
        [breakpointMediaQueries.xl]: {
          fontSize: '1.5rem',
        },
      },
      fontSize: 'calc(1.275rem + .3vw)',
    },
  ],
  5: [headingBaseStyle, { fontSize: '1.25rem' }],
  6: [headingBaseStyle, { fontSize: '1rem' }],
});
