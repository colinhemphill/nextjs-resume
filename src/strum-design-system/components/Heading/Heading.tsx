import React, { PropsWithChildren } from 'react';
import { Atoms } from '../../sprinkles.css';
import { composeWithAtoms } from '../../utils/compose';
import { headingStyleVariants } from './Heading.css';

interface HeadingBase {
  atoms?: Atoms;
  level: 1 | 2 | 3 | 4 | 5 | 6;
  text: string;
}

type Heading = RequireOnlyOne<
  PropsWithChildren<HeadingBase>,
  'text' | 'children'
>;

const Heading: React.FC<Heading> = (props) => {
  const { atoms: atomicProperties, children, level, text } = props;

  const style = composeWithAtoms(atomicProperties, headingStyleVariants[level]);

  switch (level) {
    case 1:
      return <h1 className={style}>{text || children}</h1>;
    case 2:
      return <h2 className={style}>{text || children}</h2>;
    case 3:
      return <h3 className={style}>{text || children}</h3>;
    case 4:
      return <h4 className={style}>{text || children}</h4>;
    case 5:
      return <h5 className={style}>{text || children}</h5>;
    case 6:
      return <h6 className={style}>{text || children}</h6>;
  }
};

export default Heading;
