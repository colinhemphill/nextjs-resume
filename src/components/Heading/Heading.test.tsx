import { render, screen } from '@test-utils';
import { describe, expect, test } from 'vitest';
import { Heading } from './Heading';

describe('<Heading />', () => {
  test('Snapshot', () => {
    const { container } = render(
      <>
        <Heading color="neutral" level={1}>
          Heading 1
        </Heading>
        <Heading color="neutralSubtle" level={2}>
          Heading 2
        </Heading>
        <Heading color="primary" level={3}>
          Heading 3
        </Heading>
        <Heading color="danger" level={4}>
          Heading 4
        </Heading>
        <Heading level={5}>Heading 5</Heading>
      </>,
    );
    expect(container).toMatchSnapshot();
  });

  test('Renders an accessible heading', () => {
    render(<Heading level={1}>Level 1 Heading</Heading>);

    const heading = screen.getByRole('heading', {
      level: 1,
      name: /Level 1 Heading/i,
    });

    expect(heading).toBeDefined();
  });
});
