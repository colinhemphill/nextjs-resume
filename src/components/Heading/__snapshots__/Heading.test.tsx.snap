// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Heading /> > Snapshot 1`] = `
<div>
  <h1
    class="font-bold text-neutral-12 text-4xl md:text-5xl"
  >
    Heading 1
  </h1>
  <h2
    class="font-bold text-neutral-11 text-2xl md:text-3xl"
  >
    Heading 2
  </h2>
  <h3
    class="font-bold text-primary-11 text-xl md:text-2xl"
  >
    Heading 3
  </h3>
  <h4
    class="font-bold text-danger-11 text-lg md:text-xl"
  >
    Heading 4
  </h4>
  <h5
    class="font-bold text-neutral-12 text-base md:text-lg"
  >
    Heading 5
  </h5>
</div>
`;
